MM-11102 - Added code to hyperlink channels in HTML email

waseem18 opened the pull request:
#### Summary
1. Resolves #9399
2. Added code to hyperlink channels mentioned in the message in HTML email

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-11102

waseem18 commented:
@hanzei Please review this

hanzei commented:
Thank you very much for the PR @waseem18!

@jasonblais Can you please assign a PM for review? I'v not sure who should take care of this.

hanzei commented:
@esethna Can you please take a look at this?

waseem18 commented:
@esethna Thanks for the review.

Will address that.

waseem18 commented:
Addressed the feedback in the new commit.

Email now looks like this. It now shows display name instead of channel URL handler.
<img width="807" alt="screen shot 2019-02-01 at 10 48 03 pm" src="https://user-images.githubusercontent.com/5843007/52138784-304e2400-2674-11e9-9df4-fa7ca78b310c.png">

waseem18 commented:
Any issues with Jenkins? I see the build hasn't yet started.

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

hanzei commented:
The test server have to be set up manually. Should be updated soon.

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-0df7c39f94e64806b.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Instance ID: i-0df7c39f94e64806b

esethna commented:
@cpanato same issue here, getting 502 bad gateway on the spinmint.

hanzei commented:
FYI Works for me

hanzei commented:
Same for me. Maybe we have to wait until @cpanato is back.

waseem18 commented:
Build started failing for some reason. Apparently a test case seems to be failing.

mattermod commented:
Spinmint test running for more than 7 days. This test server was terminated.

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-06c038d0bdb20a36f.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Instance ID: i-06c038d0bdb20a36f

hanzei commented:
The spinmint works now. @esethna can you review again please?

mattermod commented:
Spinmint test running for more than 7 days. This test server was terminated.

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-0f9fb0cdfc5a0418a.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Instance ID: i-0f9fb0cdfc5a0418a

mattermod commented:
Spinmint test server destroyed

jespino commented on app/notification_email.go:
```diff
@@ -284,6 +288,23 @@ func getFormattedPostTime(user *model.User, post *model.Post, useMilitaryTime bo
 }
 
+func (a *App) generateHyperlinkForChannels(postMessage, teamName, teamURL string) string {
+	r := regexp.MustCompile(`\~[-a-zA-Z0-9]+\b`)
```
I don't love the regular expressions as a generic solution to find the channel strings, that string can match with other things, but because after that the channel is search and only modify the message when that happens, I think we are safe for a lot of corner cases.

I would add one check here: the `GetMessageForNotification` generate the message for "attachment-only" posts, would be nice to check it here before we try to do do any replacement. for not replace when is an attachment only notification.

hmhealey replied:
This is related to @jespino's comment about using `GetMessageForNotification`, but this is going to cause conflicts with a few different Markdown things. The two that I can think of are:
1. Channel links inside of code blocks will be rendered as actual links
2. Strikethrough text such as ~~this~~ (`~~this~~`) may be detected as potential channel links

Since we don't support other Markdown in these emails yet, this might be fine for now, but I think it's something to be aware of.

hmhealey commented on app/notification_email.go:
```diff
@@ -284,6 +288,23 @@ func getFormattedPostTime(user *model.User, post *model.Post, useMilitaryTime bo
+	matches := r.FindAllString(postMessage, -1)
+	var channelURL, channelHyperLink string
+	for _, channelName := range matches {
+		ch, err := a.GetChannelByNameForTeamName(channelName[1:], teamName, true)
```
This is going to leak some data about whether or not a given channel name exists. For example, if I could post a link to `~firing-joe`, and if it is turned into a link, then I know that there's a channel about firing Joe even if it's private and I can't see it myself.

As far as I know, our stance on that is that this is confidential data, so we might need some logic here to not link private channels, although that would really complicate this since then the email text would be based off of who is receiving the email which would heavily change this code. I'm going to bring this up with our security team to get a firm decision on this since while this has been our stance in the past, there are other ways to get this information.

hmhealey replied:
We're going to have to change how this works since we do consider channel names confidential, and we can't customize the email for different users with how email notifications are set up.

@esethna How would you feel if we just turn anything that looks like a channel link into a link? It's obviously not ideal since a user might make a broken link by accident or if they make a typo, but it would at least fix the issue that those links aren't clickable. Alternatively, we could only link public channels in emails which fixes the problem of broken links, but it prevents private channel links from working.

esethna replied:
Sorry, missed this notification.

@hmhealey @waseem18 given the above security concern I think **we should only be linking Public channels**. Reason being that if we link everything that looks like a channel name we might link things such as `~3 hours from now` or `around ~thursday evening`

waseem18 replied:
Sure @esethna

hmhealey commented on app/notification_email.go:
```diff
@@ -284,6 +288,23 @@ func getFormattedPostTime(user *model.User, post *model.Post, useMilitaryTime bo
+	for _, channelName := range matches {
+		ch, err := a.GetChannelByNameForTeamName(channelName[1:], teamName, true)
+		if ch == nil || err != nil {
+			mlog.Warn(fmt.Sprintf("Encountered error while generating hyperlink for %s, as channel with that name doesn't exist.", channelName[1:]))
```
Can you downgrade this to an `mlog.Debug`? This is something that can happen normally if a user mistypes a channel name or uses one of the Markdown things I mentioned above, so I don't think it needs to be a warning message that might clutter up the logs.

waseem18 commented:
I'll address the reviews soon.

mattermod commented:
This issue has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

/cc @jasonblais @hanzei

jasonblais commented:
Hey @waseem18! I know you've been working on the Jenkins plugin a ton recently. Is this something you'd be open to helping take a look at after the Jenkins plugin is ready?

waseem18 commented:
@jasonblais I'm about to push an initial release for the Jenkins plugin. As it's Friday today, I'll try to wrap this up over this weekend.

hanzei commented:
Hey @waseem18,

Since the work on the jenkins plugin can be mostly completed, do you have the time to finish this PR?

Please note that we will close this PR after another 10 days of inactivity. If you don't object, other contributors might complete this PR for you.

hanzei commented:
This PR has become orphaned and will be closed. Still, thanks for your contribution @waseem18
